src/script/descriptor.cpp

```diff
@@ -1414,34 +1414,76 @@ enum class ParseScriptContext {
- * @param[out] out the key path
+ * @param[out] out Vector of parsed key paths
  * @param[out] apostrophe only updated if hardened derivation is found
  * @param[out] error parsing error message
+ * @param[in] allow_multipath Allows the parsed path to use the multipath specifier
  * @returns false if parsing failed
  **/
-[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out, bool& apostrophe, std::string& error)
+[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, std::vector<KeyPath>& out, bool& apostrophe, std::string& error, bool allow_multipath = false)
 {
+    out.emplace_back();
+
     for (size_t i = 1; i < split.size(); ++i) {
-        Span<const char> elem = split[i];
-        bool hardened = false;
-        if (elem.size() > 0) {
-            const char last = elem[elem.size() - 1];
-            if (last == '\'' || last == 'h') {
-                elem = elem.first(elem.size() - 1);
-                hardened = true;
-                apostrophe = last == '\'';
+        std::vector<Span<const char>> to_process;
+        Span<const char> item = split[i];
+        if (item.size() > 0 && item[0] == '<' && item[item.size() - 1] == '>') {
+            if (!allow_multipath) {
+                error = strprintf("Key path value '%s' specifies multipath in a section where multipath is not allowed", std::string(item.begin(), item.end()));
+                return false;
+            }
+            if (out.size() > 1) {
+                error = "Multiple multipath key path specifiers found";
+                return false;
+            }
+            to_process = Split(item.subspan(1, item.size() - 2), ';');
+            if (to_process.size() <= 1) {
+                error = "Multipath key path specifiers must have at least two items";
+                return false;
+            }
+        } else {
+            to_process.push_back(item);
+        }
+        std::vector<uint32_t> multipath_elems;
+        for (auto& elem : to_process) {
+            bool hardened = false;
+            if (elem.size() > 0) {
+                const char last = elem[elem.size() - 1];
+                if (last == '\'' || last == 'h') {
+                    elem = elem.first(elem.size() - 1);
+                    hardened = true;
+                    apostrophe = last == '\'';
+                }
+            }
+            uint32_t p;
+            if (!ParseUInt32(std::string(elem.begin(), elem.end()), &p)) {
+                error = strprintf("Key path value '%s' is not a valid uint32", std::string(elem.begin(), elem.end()));
+                return false;
+            } else if (p > 0x7FFFFFFFUL) {
+                error = strprintf("Key path value %u is out of range", p);
+                return false;
+            }
+            p |= ((uint32_t)hardened) << 31;
+            if (std::find(multipath_elems.begin(), multipath_elems.end(), p) != multipath_elems.end()) {
```

Comment:
We need the vector since order has to be preserved, and since the number of elements should be small, I don't think the performance impact is significant. Also, `std::find` is `O(n)`.

Reply:
It's `O(n)` but it's called in a loop over `n`.. So it effectively does `(n^2 - n)/2` operations, which is `O(n^2)`.